prov/efa/src/rdm/efa_rdm_ep_fiops.c

```diff
@@ -982,6 +1056,71 @@ void efa_rdm_ep_update_shm(struct efa_rdm_ep *ep)
 		efa_rdm_ep_close_shm_resources(ep);
 }
 
+static inline
+int efa_rdm_ep_insert_cntr_ibv_cq_poll_list(struct efa_rdm_ep *ep)
+{
+	int i, ret;
+	struct efa_cntr *efa_cntr;
+	struct util_cntr *util_cntr;
+	struct efa_rdm_cq *tx_cq, *rx_cq;
+	tx_cq = ep->tx_cq;
+	rx_cq = ep->rx_cq;
+
+	for (i=0; i<CNTR_CNT; i++) {
+		util_cntr = ep->base_ep.util_ep.cntrs[i];
+		if (util_cntr) {
+			efa_cntr = container_of(util_cntr, struct efa_cntr, util_cntr);
+			if (tx_cq) {
+				ret = efa_ibv_cq_poll_list_insert(&efa_cntr->ibv_cq_poll_list, &efa_cntr->util_cntr.ep_list_lock, &tx_cq->ibv_cq);
+				if (ret)
+					return ret;
+			}
+			if (rx_cq) {
+				ret = efa_ibv_cq_poll_list_insert(&efa_cntr->ibv_cq_poll_list, &efa_cntr->util_cntr.ep_list_lock, &rx_cq->ibv_cq);
+				if (ret)
+					return ret;
+			}
+		}
+	}
+
+	return FI_SUCCESS;
+}
+
+static inline
+int efa_rdm_ep_insert_cq_ibv_cq_poll_list(struct efa_rdm_ep *ep)
+{
+	int ret;
+	struct efa_rdm_cq *tx_cq, *rx_cq;
+	/* cross referencing */
+	tx_cq = ep->tx_cq;
+	rx_cq = ep->rx_cq;
+
+	if (tx_cq) {
+		ret = efa_ibv_cq_poll_list_insert(&tx_cq->ibv_cq_poll_list, &tx_cq->util_cq.ep_list_lock, &tx_cq->ibv_cq);
+		if (ret)
+			return ret;
+
+		if (rx_cq) {
+			ret = efa_ibv_cq_poll_list_insert(&tx_cq->ibv_cq_poll_list, &tx_cq->util_cq.ep_list_lock, &rx_cq->ibv_cq);
+			if (ret)
+				return ret;
```

Comment:
Same as above. This looks odd...

Reply:
Same, you need to a cross referencing. Application needs to poll both tx and rx ibv cq even if it only read OFI tx cq. This complexity will be removed in OFI 2.0, which proposes to only allow bind 1 cq per endpoint for both tx and rx operations